Wire Edit Colony Flow

rdig opened the pull request:
## Description

This PR wires in the colony edit action flow:
- the actual colony action creation (via the modal)
- refactor to the way we upload avatars
- wire in the actual action for the action page
- wire in the subgraph action for this
- refactor colony data to come from subgraph events
- refactor the create colony saga to send metadata

**Changes**
- [x] Added `EDIT_COLONY` redux action types
- [x] Added `editColonyAction` actions saga
- [x] Update `EditColonyDialog` to send in correct values
- [x] Added colony edit action and event copy
- [x] Refactored colony action resolver to return "Wrong Colony" action type -- _it's a check to prevent opening actions from colonies that they not belong to_
- [x] `ActionsPage` now properly displays colony edit action
- [x] `DisplayWidget` now properly displays colony edit action
- [x] Fetch IPFS metadata in the actions resolver
- [x] Added colony metadata event from subgraph -- to populate the actions list
- [x] Fetch the colony's data from the subgraph
- [x] Update the cache after joining / leaving a colony
- [x] Subscribe to colony after creating it
- [x] Fix user token minting permissions refresh
- [x] Detect type of colony metadata change
- [x] Wire up `TokenManagement` dialog
- [x] Refactored `FriedlyName`
- [x] Cleanup old admin domains / profile logic
- [x] Better token validation and fail safes
- [x] **Many** improvements to local cache updates

Also note that:
- this uses JoinColony/subgraph#11 to update the subgraph colony metadata history part
- this uses JoinColony/colonyServer#132 where it removes the domains and colonies

So you'll need to run `npm i` and `npm provision` in order to make this work

**Demo**

Edit colony title

![demo-colony-edit-action](https://user-images.githubusercontent.com/1193222/105093014-b99dd180-5aaa-11eb-8d91-4af87372711d.gif)

Edit colony avatar

![demo-colony-edit-logo](https://user-images.githubusercontent.com/1193222/105253806-e58f8480-5b88-11eb-8ac4-97b6e6ebe3de.gif)

Edit tokens
![demo-manage-tokens](https://user-images.githubusercontent.com/1193222/105630781-b9f3ef00-5e53-11eb-8099-716481eece8e.gif)

Subscribe / unsubscribe
![demo-subscriptions](https://user-images.githubusercontent.com/1193222/105630789-c2e4c080-5e53-11eb-9763-d416bf380ef0.gif)


Resolves DEV-77
Resolves DEV-140
Resolves DEV-144
Resolves DEV-147
Resolves DEV-156

And a bunch of in-passing fixes / cleanup

Resolves DEV-34
Resolves DEV-35
Resolves DEV-57
Resolves DEV-108
Resolves DEV-111

rdig commented:
Sorry about the size of this :(

But on the plus side, it ties everything up in a neat little bow. :bowtie:

alicjakujawa commented on src/modules/dashboard/components/EditColonyDetailsDialog/EditColonyDetailsDialogForm.tsx:
```diff
@@ -197,7 +258,12 @@ const EditColonyDetailsDialogForm = ({
+            !userHasPermission ||
+            !isValid ||
+            avatarFileError ||
+            !canValuesBeUpdate()
```
why this is function? 🤔

rdig replied:
Custom validation, as we don't have a proper way (using `yup`) of checking if the name _or_ logo changed, but didn't change to the same values they were initially.

It's not ideal. but It's the only way we currently can validate this form properly

alicjakujawa replied:
Hmmm, so why not make it just a variable?

rdig replied:
Actually a good point.

So initially, this had more "complex" logic and I needed a function to achieve that. But looking at the final result, it is indeed a good candidate to just change into a variable

rdig replied:
I thought I did? Did I not commit the change? What's going on here????

alicjakujawa commented on src/modules/dashboard/sagas/actions.ts:
```diff
@@ -735,6 +739,210 @@ function* createDomainAction({
+     * Validate the required values for the payment
+     */
+    if (!colonyDisplayName && colonyDisplayName !== null) {
+      throw new Error('A colony name is required in order to edit the colony');
```
why !colonyDisplayName && colonyDisplayName !== null instead of just !colonyDisplayName ??

rdig replied:
Because if it's undefined I don't want to send the date, but if it's null, I want to (this means the colony display name was removed by the user)

Same goes for the avatar

alicjakujawa commented on src/modules/dashboard/sagas/utils/updateColonyDisplayCache.ts:
```diff
@@ -0,0 +1,59 @@
+      // eslint-disable-next-line consistent-return
+      displayName: () => {
+        if (displayName || displayName === null) {
+          return displayName;
```
im lost here a bit. So if we want return displayName always (when its null or not) then why we just not do
displayName: () => displayName ?

rdig replied:
No. We don't want to return the `undefined` case, we just want either a string value, or null.

So that's why we have to do this little dance of filtering out null :(

alicjakujawa commented on src/utils/events/index.ts:
```diff
@@ -270,6 +284,44 @@ const getCreateDomainActionValues = async (
+const getColonyEditActionValues = async (
+  processedEvents: ProcessedEvent[],
+): Promise<Partial<ActionValues>> => {
+  const mintTokensEvent = processedEvents.find(
```
why this name? 🤔

rdig replied:
I guess I copy-pasted it from somewhere :)

I'll change it.

rdig commented on src/data/graphql/typeDefs.ts:
```diff
@@ -38,6 +38,9 @@ export default gql`
     annotationHash: String
+    colonyDisplayName: String!
+    colonyAvatarHash: String!
+    colonyTokens: [String]!
```
It might be an empty array

alicjakujawa replied:
From my understanding, by adding String! inside of array, you just forcing, that if array has item, then it need to be string. so with  [String!]! ->
[] - is ok
['string'] - is ok
[null] - is not ok

rdig commented on src/data/resolvers/colony.ts:
```diff
@@ -138,60 +300,66 @@ export const colonyResolvers = ({
+             * Converting this to string basically converts the error object into
+             * a string form
+             */
+            address: address.toString().toLowerCase(),
```
It won't be.

This comes from colonyName query which takes the colony name and converts it into an address.

If that fails this doesn't get called.

rdig commented on src/data/resolvers/user.ts:
```diff
@@ -155,5 +163,45 @@ export const userResolvers = ({
+            if (subscribedColony) {
+              userColonies.push(subscribedColony);
+            }
+            return null;
```
I don't understand how that would be more beneficial here ?

Sure it's a different approach to this sure, and it would remove the need for returns, but I don't see the benefit of using this over my current implementation.

alicjakujawa replied:
So "map" is used to "decorate" values in the array, not to loop over array and create different array

rdig replied:
I mean, I've been using `map` to modify an array my whole life :)

But I guess what you're saying makes sense, so I'll make the change.

rdig commented on src/modules/core/components/TokenEditDialog/TokenEditDialog.tsx:
```diff
@@ -90,37 +105,71 @@ const TokenEditDialog = ({
+            .filter((address) => {
+              if (address === AddressZero) {
+                return false;
+              }
```
Sure. Initially I thought my two filters would look "cleaner" :)

chinins commented on src/modules/core/components/Fields/Input/InputComponent.tsx:
```diff
@@ -111,6 +112,8 @@ const InputComponent = ({
         placeholder={placeholder}
         ref={innerRef}
         maxLength={maxLength}
+        // value={value}
```
should the comment be deleted or uncommented?

chinins commented on src/modules/core/components/FriendlyName/FriendlyName.tsx:
```diff
@@ -30,20 +31,42 @@ interface Props {
+    }
+  }
+
+  // const isColonyAddress = colony && walletAddress === colony.colonyAddress;
```
should the commented lines be deleted?

chinins commented on src/modules/core/components/TokenEditDialog/TokenEditDialog.tsx:
```diff
@@ -1,19 +1,26 @@
 import React, { useCallback, useState, useMemo } from 'react';
```
Couple of points regarding token management dialog:

1. When there is an error in a new address (in the input) the color doesn't change to red. I think to be consistent with other dialogs/forms it should change color.

2. When deleting token there are 2 warning messages for Appolo cache (i saw it only when deleting):

```
warning 1:
Cache data may be lost when replacing the tokenAddresses field of a ProcessedColony object.

To address this problem (which is not a bug in Apollo Client), define a custom merge function for the ProcessedColony.tokenAddresses field, so InMemoryCache can safely merge these objects:

  existing: ["0x8972e86549bb8E350673e0562fba9a4889d01637","0x7F176F534C31294fDbA77D80bB9d01279820C063"]
  incoming: ["0x7F176F534C31294fDbA77D80bB9d01279820C063"]

For more information about these options, please refer to the documentation:

  * Ensuring entity objects have IDs: https://go.apollo.dev/c/generating-unique-identifiers
  * Defining custom merge functions: https://go.apollo.dev/c/merging-non-normalized-objects

  
  warning 2:
  Cache data may be lost when replacing the tokens field of a ProcessedColony object.

To address this problem (which is not a bug in Apollo Client), define a custom merge function for the ProcessedColony.tokens field, so InMemoryCache can safely merge these objects:

  existing: [{"__ref":"Token:0x0000000000000000000000000000000000000000"},{"__ref":"Token:0x8972e86549bb8E350673e0562fba9a4889d01637"},{"__ref":"Token:0x7F176F534C31294fDbA77D80bB9d01279820C063"}]
  incoming: [{"__ref":"Token:0x0000000000000000000000000000000000000000"},{"__ref":"Token:0x7F176F534C31294fDbA77D80bB9d01279820C063"}]

For more information about these options, please refer to the documentation
...
```
 

ArmandoGraterol reviewed (approved):
It's working great for me :ok_hand: This a beast of PR

alicjakujawa reviewed (changes requested):
<img width="755" alt="Screenshot 2021-01-27 at 12 13 21" src="https://user-images.githubusercontent.com/13069555/105983750-77bbef00-6099-11eb-8302-419585149264.png">
* We need to handle error when user actually will choose file >1MB
* Also i think, im not able to choose svg format avatar
Apart of this bug just update graphql schema type and this HUGE PR will be ready to merge 🔥

rdig commented:
@alicjakujawa 

> * We need to handle error when user actually will choose file >1MB * Also i think, im not able to choose svg format avatar Apart of this bug just update graphql schema type and this HUGE PR will be ready to merge fire

re-install node modules and try this again please. We handle upload errors. I think that's coming from a library not existing on your envrionment (I've had this before when testing Olga's PR)

alicjakujawa commented:
@rdig removed and installed node_modules, and same error. Also im not able to upload svg file

chinins reviewed (approved):
Looks and works fine for me.

rdig commented:
@alicjakujawa 

1. Cannot replicate your first issue, with not handling errors. Works fine locally and even on QA. Here's a demo:
![demo-cannot-upload-large-files](https://user-images.githubusercontent.com/1193222/106046660-394f2000-60eb-11eb-99d3-d2c9e96dfad9.gif)

2. You were right about the svg files, the mime type was not set correctly. I've fixed that now:
![demo-can-upload-svg](https://user-images.githubusercontent.com/1193222/106046701-4704a580-60eb-11eb-9ddd-94760bfddfe6.gif)

3. I've changed the `canValuesBeUpdate` function into a simple variable now

alicjakujawa reviewed (approved):
🔥
